# PJ01 - Choose Your Own Analysis

In [36]:
__author__ = "730395347"

# Hypothesis: Are students who come to office hours more likely to recommend the course?

In this case, I need to compare the course recommendations of students who don't attend office hours and the course recommendations of students who do attend office hours.

## Prediction

I think that students who go to office hours in general are more likely to recommend the course. 

In [37]:
%reload_ext autoreload
%autoreload 2

First I will be changing the data into a column-oriented table so it will be easier to find the values associated with course recommendations and the students who have attended office hours sections. I also think it would be interesting to see how often the average student in this course attends office hours and to see if the average student would recommend this course. To find this, it would be easier to put the data into a column-oriented table. 

In [38]:
DATA_DIRECTORY="../../data"
DATA_FILE_PATH=f"{DATA_DIRECTORY}/survey.csv"

The purpose of the code cell above is create a reference to the survey data set that will be used throughout the rest of the code in the notebook. 

In [39]:
from data_utils import read_csv_rows
data_rows: list[dict[str, str]] = read_csv_rows(DATA_FILE_PATH)

if len(data_rows) == 0:
    print("Go implement read_csv_rows in data_utils.py")
    print("Be sure to save your work before re-evaluating this cell!")
else:
    print(f"Data File Read: {DATA_FILE_PATH}")
    print(f"{len(data_rows)} rows")
    print(f"{len(data_rows[0].keys())} columns")
    print(f"Columns names: {data_rows[0].keys()}")

Data File Read: ../../data/survey.csv
573 rows
36 columns
Columns names: dict_keys(['row_number', 'year', 'unc_status', 'comp_major', 'primary_major', 'data_science', 'prereqs', 'prior_exp', 'AP_Principles', 'AP_A', 'other_comp', 'prior_time', 'languages', 'section', 'in_person_classes', 'hours_online_social', 'hours_online_work', 'lesson_time', 'sync_perf', 'all_sync', 'own_notes', 'own_examples', 'oh_visits', 'ls_effective', 'lsqs_effective', 'programming_effective', 'qz_effective', 'oh_effective', 'tutoring_effective', 'kaki_effective', 'pace', 'difficulty', 'understanding', 'interesting', 'valuable', 'would_recommend'])


The purpose of the code cell above is to implement the utility functions that will read the CSV file.

In [40]:
from data_utils import columnar

data_cols: dict[str, list[str]] = columnar(data_rows)

if len(data_cols.keys()) == 0:
    print("Complete your implementation of columnar in data_utils.py")
    print("Be sure to follow the guidelines above and save your work before re-evaluating!")
else:
    print(f"{len(data_cols.keys())} columns")
    print(f"Columns names: {data_cols.keys()}")

36 columns
Columns names: dict_keys(['row_number', 'year', 'unc_status', 'comp_major', 'primary_major', 'data_science', 'prereqs', 'prior_exp', 'AP_Principles', 'AP_A', 'other_comp', 'prior_time', 'languages', 'section', 'in_person_classes', 'hours_online_social', 'hours_online_work', 'lesson_time', 'sync_perf', 'all_sync', 'own_notes', 'own_examples', 'oh_visits', 'ls_effective', 'lsqs_effective', 'programming_effective', 'qz_effective', 'oh_effective', 'tutoring_effective', 'kaki_effective', 'pace', 'difficulty', 'understanding', 'interesting', 'valuable', 'would_recommend'])


The purpose of the cell above is to take a dictionary that is row-oriented and change it into a column-oriented dictionary. It is a lot easier to find values that are associated with a particular column. Now we find what variables to look for. 

In [41]:
data_cols_head: dict[str, list[str]] = head(data_cols, 5)

tabulate(data_cols_head, data_cols_head.keys(), "html")

'<table>\n<thead>\n<tr><th style="text-align: right;">  row_number</th><th style="text-align: right;">  year</th><th>unc_status           </th><th>comp_major  </th><th>primary_major                   </th><th>data_science  </th><th>prereqs                                          </th><th>prior_exp                   </th><th>AP_Principles  </th><th>AP_A  </th><th>other_comp  </th><th>prior_time  </th><th>languages  </th><th>section  </th><th>in_person_classes  </th><th>hours_online_social  </th><th>hours_online_work  </th><th style="text-align: right;">  lesson_time</th><th style="text-align: right;">  sync_perf</th><th style="text-align: right;">  all_sync</th><th style="text-align: right;">  own_notes</th><th style="text-align: right;">  own_examples</th><th style="text-align: right;">  oh_visits</th><th style="text-align: right;">  ls_effective</th><th style="text-align: right;">  lsqs_effective</th><th style="text-align: right;">  programming_effective</th><th style="text-align: right;">  qz_effective</th><th>oh_effective  </th><th>tutoring_effective  </th><th>kaki_effective  </th><th style="text-align: right;">  pace</th><th style="text-align: right;">  difficulty</th><th style="text-align: right;">  understanding</th><th style="text-align: right;">  interesting</th><th style="text-align: right;">  valuable</th><th style="text-align: right;">  would_recommend</th></tr>\n</thead>\n<tbody>\n<tr><td style="text-align: right;">           0</td><td style="text-align: right;">    23</td><td>Returning UNC Student</td><td>No          </td><td>Advertising and Public Relations</td><td>No            </td><td>MATH 129P, MATH 130                              </td><td>None to less than one month!</td><td>               </td><td>      </td><td>            </td><td>            </td><td>           </td><td>In-Person</td><td>                   </td><td>3 to 5 hours         </td><td>3 to 5 hours       </td><td style="text-align: right;">            5</td><td style="text-align: right;">          2</td><td style="text-align: right;">         2</td><td style="text-align: right;">          4</td><td style="text-align: right;">             2</td><td style="text-align: right;">          2</td><td style="text-align: right;">             7</td><td style="text-align: right;">               6</td><td style="text-align: right;">                      6</td><td style="text-align: right;">             5</td><td>5             </td><td>                    </td><td>3               </td><td style="text-align: right;">     5</td><td style="text-align: right;">           5</td><td style="text-align: right;">              4</td><td style="text-align: right;">            5</td><td style="text-align: right;">         6</td><td style="text-align: right;">                7</td></tr>\n<tr><td style="text-align: right;">           1</td><td style="text-align: right;">    24</td><td>Returning UNC Student</td><td>No          </td><td>Anthropology                    </td><td>No            </td><td>MATH 129P, MATH 231                              </td><td>None to less than one month!</td><td>               </td><td>      </td><td>            </td><td>            </td><td>           </td><td>In-Person</td><td>                   </td><td>5 to 10 hours        </td><td>5 to 10 hours      </td><td style="text-align: right;">            4</td><td style="text-align: right;">          5</td><td style="text-align: right;">         4</td><td style="text-align: right;">          7</td><td style="text-align: right;">             6</td><td style="text-align: right;">          0</td><td style="text-align: right;">             6</td><td style="text-align: right;">               5</td><td style="text-align: right;">                      7</td><td style="text-align: right;">             5</td><td>              </td><td>                    </td><td>6               </td><td style="text-align: right;">     5</td><td style="text-align: right;">           3</td><td style="text-align: right;">

In the code above, I created a table showing the first five responses so I could find what headers were the most important to look at. I settled on 'oh_visits' and 'would_recommend' as I thought it was most relevant to the hypothesis. 

## A Look Into The Hypothesis

To get a feel for if my hypothesis is correct or not, I have to look at the values associated with 'oh_visits' and 'would_recommend'. Since the data set is so large, I will be looking at the first 20 rows. I will also be looking at how many responses are in each section as I am curious to see how many students have put down an answer.

In [42]:
from data_utils import head
from data_utils import select
from tabulate import tabulate

selected_data: dict[str, list[str]] = select(data_cols, ["oh_visits", "would_recommend"])

tabulate(head(selected_data, 20), selected_data.keys(), "html")


'<table>\n<thead>\n<tr><th style="text-align: right;">  oh_visits</th><th style="text-align: right;">  would_recommend</th></tr>\n</thead>\n<tbody>\n<tr><td style="text-align: right;">          2</td><td style="text-align: right;">                7</td></tr>\n<tr><td style="text-align: right;">          0</td><td style="text-align: right;">                7</td></tr>\n<tr><td style="text-align: right;">          0</td><td style="text-align: right;">                7</td></tr>\n<tr><td style="text-align: right;">          4</td><td style="text-align: right;">                7</td></tr>\n<tr><td style="text-align: right;">          1</td><td style="text-align: right;">                7</td></tr>\n<tr><td style="text-align: right;">          1</td><td style="text-align: right;">                4</td></tr>\n<tr><td style="text-align: right;">          0</td><td style="text-align: right;">                6</td></tr>\n<tr><td style="text-align: right;">          1</td><td style="text-align: right;">                4</td></tr>\n<tr><td style="text-align: right;">          1</td><td style="text-align: right;">                6</td></tr>\n<tr><td style="text-align: right;">          2</td><td style="text-align: right;">                5</td></tr>\n<tr><td style="text-align: right;">          3</td><td style="text-align: right;">                4</td></tr>\n<tr><td style="text-align: right;">          2</td><td style="text-align: right;">                4</td></tr>\n<tr><td style="text-align: right;">          0</td><td style="text-align: right;">                5</td></tr>\n<tr><td style="text-align: right;">          3</td><td style="text-align: right;">                6</td></tr>\n<tr><td style="text-align: right;">          1</td><td style="text-align: right;">                6</td></tr>\n<tr><td style="text-align: right;">          0</td><td style="text-align: right;">                6</td></tr>\n<tr><td style="text-align: right;">          1</td><td style="text-align: right;">                7</td></tr>\n<tr><td style="text-align: right;">          2</td><td style="text-align: right;">                5</td></tr>\n<tr><td style="text-align: right;">          1</td><td style="text-align: right;">                1</td></tr>\n<tr><td style="text-align: right;">          2</td><td style="text-align: right;">                3</td></tr>\n</tbody>\n</table>'

In the code cell above, I used both the head and select functions to find the first 20 responses for 'oh_visits' and 'would_recommend'. Based on the the responses, it is hard to tell if there is an association between students going to office hours and students recommending the class. 

In [43]:
from data_utils import count

oh_counts: dict[str, int] = count(selected_data["oh_visits"])
print(f"Break down of students responses on how often they visit office hours: {oh_counts}")

recommend_counts: dict[str, int] = count(selected_data["would_recommend"])
print(f"Break down of student responses on if they recommend the course: {recommend_counts}")

Break down of students responses on how often they visit office hours: {'2': 117, '0': 185, '4': 19, '1': 185, '3': 44, '5': 23}
Break down of student responses on if they recommend the course: {'7': 197, '4': 88, '6': 134, '5': 92, '1': 12, '3': 34, '2': 16}


The code cell above shows us the breakdown of each response which will is helpful, but for now we need to find the total of all the responses in 'oh_visits' and 'would_recommend'. We could count all the numbers up but it would be more effective to create a helper function. 

In [44]:
oh_counts: dict[str, int] = count(selected_data["oh_visits"])
recommend_counts: dict[str, int] = count(selected_data["would_recommend"])

def sum (data: dict[str, int]) -> int:
    """Adds up all the responses in a column oriented dictionary."""
    result: int = 0
    for value in data:
        result += data[value]
    return result
    
oh: int = sum(oh_counts)
rec: int = sum(recommend_counts)
print(f"Total number of students who responsed to how often they visit office hours: {oh}")
print(f"Total number of students who responded to if they would recommend the course: {rec}")

Total number of students who responsed to how often they visit office hours: 573
Total number of students who responded to if they would recommend the course: 573


With the helper function we were able to find the total number of students who responded. It is good that both of the values yield the same result (573 responses). This means that the data for both variables is the same size and it is much easier to work with data that is the same size. 

In [45]:
def oh_often(data: dict[str, int]) -> int:
    result: int = 0
    for value in data:
        if value == '5':
            result += data[value]
        if value == '4':
            result += data[value]
        if value == '3':
            result += data[value]
        if value == '2':
            result 
        if value == '1':
            result 
        if value == '0':
            result
    return result 

def most_rec(data: dict[str, int]) -> int:
    result: int = 0
    for value in data:
        if value == '7':
            result += data[value]
        if value == '6':
            result += data[value]
        if value == '5':
            result += data[value]
        if value == '4':
            result 
        if value == '3':
            result 
        if value == '2':
            result
        if value == '1':
            result
        if value == '0':
            result
    return result 

print(f"Total number of students who visit office hours the most (response 3 to 5): {oh_often(oh_counts)}")
print(f"Total number of students who recommend the course the most (response 5 to 7): {most_rec(recommend_counts)}")

Total number of students who visit office hours the most (response 3 to 5): 86
Total number of students who recommend the course the most (response 5 to 7): 423


With the helper function above, I wanted to compare the number of students who recommend the course (response 5 to 7) with students who visit office hours the most (response 3 to 5). From the results you can see that most of the students who recommend the course the most actually do not attend office hours frequently. 

In [46]:
def oh_often1(data: dict[str, int]) -> int:
    result: int = 0
    for value in data:
        if value == '5':
            result += data[value]
        if value == '4':
            result += data[value]
        if value == '3':
            result += data[value]
        if value == '2':
            result += data[value]
        if value == '1':
            result += data[value]
        if value == '0':
            result
    return result 

def most_rec1(data: dict[str, int]) -> int:
    result: int = 0
    for value in data:
        if value == '7':
            result += data[value]
        if value == '6':
            result += data[value]
        if value == '5':
            result += data[value]
        if value == '4':
            result 
        if value == '3':
            result 
        if value == '2':
            result
        if value == '1':
            result
        if value == '0':
            result
    return result 

print(f"Total number of students who visit office hours (response 1 to 5): {oh_often1(oh_counts)}")
print(f"Total number of students who recommend the course the most (response 5 to 7): {most_rec1(recommend_counts)}")

Total number of students who visit office hours (response 1 to 5): 388
Total number of students who recommend the course the most (response 5 to 7): 423


With the helper function above, I wanted to compare the number of students who recommend the course (response 5 to 7) with students who visit office hours in general (response 1 to 5). From the results you can see that the number of students who attend office hours in general is close to the number of students who recommend the course. There is difference of 35 between the two numbers. These results mean that students who visit office hours in general may be more likely to recommend the course.

# Conclusion 
My hypothesis was correct, that students who visit office hours in general are more likely to recommend the course. It is interesting though to see that many students actually do not attend a large number of office hours but many students do attend office hours one or twice. This may be why many of these students recommend the class as they find office hours accessible and helpful but the class is not too difficult for them to go to office hours multiple times. But based on the results I have calculated, casuation is only implied, the results do not mean that students who go to office hours WILL recommend the class. It only implies that there is some relation between the two variables. 